src/test/java/hudson/tasks/MailerTest.java

```diff
@@ -254,6 +258,36 @@ public void testMessageText() throws Exception {
             .buildAndCheckContent();
     }
 
+    @Issue("JENKINS-37812")
+    @Test
+    public void testFailureSendMessage() throws Exception {
+        DumbSlave node = rule.createSlave();
+        Mailer m = new MailerDisconnecting(rule, node, RECIPIENT, true, false);
+
+        new TestProject(m)
+                .withNode(node)
+                .failure()
+                .buildAndCheckSending();
+    }
+
+    @Issue("JENKINS-37812")
+    @Test
+    public void testPipelineCompatibility() throws Exception {
+        WorkflowJob p = rule.jenkins.createProject(WorkflowJob.class, "p");
+        p.setDefinition(new CpsFlowDefinition(
+                "node {\n"
+                        + "    catchError {error 'oops'}\n"
+                        + "    step([$class: 'Mailer', recipients: '" + RECIPIENT + "'])\n"
+                        + "}", true));
+
+        Mailbox inbox = getMailbox(RECIPIENT);
+        inbox.clear();
+        WorkflowRun b = rule.assertBuildStatus(Result.FAILURE, p.scheduleBuild2(0).get());
+        assertEquals(1, inbox.size());
```

Comment:
NIT: given you're already using `assertThat` below, using `hasSize` here would yield a clearer message in case of failure.